File: problem/14/python/solve_14.py
#!/usr/bin/env python


def get_collatz_length(n):
        passes = 0
        x = n
        while True:
                if x == 1:
                        return passes
                if x % 2 == 0:
                        x /= 2
                else:
                        x = 3 * x + 1
                passes += 1

results_a = []
results_b = []
for f in range(1, 1000000):
        results_a.append(f)
        results_b.append(get_collatz_length(f))

print max(results_b)
print results_a[results_b.index(max(results_b))]

File: problem/14/perl/solve_14.pl
#!/usr/bin/env perl6
# http://projecteuler.net/problem=14

use v6;

sub get_collatz_length($n is copy) {
        my $passes = 0;
        loop (;;) {
                if $n == 1 { return $passes; }
                if $n % 2 == 0 {
                        $n /= 2;
                } else {
                        $n = 3 * $n + 1
                }
                $passes++;
        }
}

my %results;
loop (my $i = 1; $i < 1000000; $i++) { %results{get_collatz_length($i)}.push($i); }

say "number: " ~ %results{[max] %results.keys>>.Int}.perl;
say "length: " ~ [max] %results.keys>>.Int;
#n → n/2 (n is even)
#n → 3n + 1 (n is odd)